File: src/main/java/com/nibado/projects/advent/Day12.java
package com.nibado.projects.advent;

import com.fasterxml.jackson.databind.JsonNode;

import java.util.Iterator;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Function;

import static com.nibado.projects.advent.Util.printAnswer;
import static com.nibado.projects.advent.Util.readResourceAsJson;

public class Day12 implements Runnable {

    @Override
    public void run() {
        JsonNode input = readResourceAsJson("/day12.txt");

        printAnswer(12, "One", sumTree(input, n -> false));
        printAnswer(12, "Two", sumTree(input, n -> hasValue(n, "red")));
    }

    public int sumTree(JsonNode node, Function<JsonNode, Boolean> filter) {
        if(node.isInt()) {
            return node.asInt();
        }
        if(filter.apply(node)) {
            return 0;
        }
        final AtomicInteger sum = new AtomicInteger(0);

        node.forEach(n -> sum.addAndGet(sumTree(n, filter)));

        return sum.get();
    }

    public boolean hasValue(JsonNode node, String value) {
        Iterator<JsonNode> iter = node.iterator();

        if(!node.isObject()) {
            return false;
        }
        while(iter.hasNext()) {
            if(value.equals(iter.next().asText())) {
                return true;
            }
        }

        return false;
    }

    public static void main(String[] argv) {
        new Day12().run();
    }
}

File: src/main/scala/com/nibado/projects/adventscala/Day12.scala
package com.nibado.projects.adventscala

import spray.json._

class Day12 extends Day {
  override def run(): Unit = {
    val json = getResourceAsJson("/day12.txt")

    printAnswer(12, "One", sumTree(json, n => false))
    printAnswer(12, "Two", sumTree(json, n => hasValue(n, "red")))
  }

  def sumTree(node: JsValue, filter: (Map[String, JsValue] => Boolean)) : Int = {
    node match {
      case JsNumber(n) => n.intValue()
      case JsArray(a) => a.map(e => sumTree(e, filter)).sum
      case JsObject(obj) => if(filter(obj)) { 0 } else obj.values.map(v => sumTree(v, filter)).sum
      case _ => 0
    }
  }

  def hasValue(obj: Map[String, JsValue], value: String) : Boolean = {
    obj.values.exists { case JsString(s) => s.equals(value) case _ => false }
  }
}
